### DISCUSSION TITLE
Validation pseudo-class variants

### DISCUSSION THREAD

**Discussion by manix** [tailwindlabs/tailwindcss#2465] [Ideas]
Hey, I'm pretty new to this, excuse me if my terminology is not on point.

I'm looking at https://tailwindcss.com/docs/pseudo-class-variants and there don't seem to be variants for `invalid` and `valid` pseudo selectors. Am I missing something? I think it would be nice if they could be added to the official library.

**Comment by simonswiss** [comment]
Hi @manix!

You are correct, there are no `invalid` and `valid` variants in Tailwind CSS.

I cannot tell you whether or not these would be added to core at some point, but you can [create your own custom variants](https://tailwindcss.com/docs/pseudo-class-variants#creating-custom-variants) with the plugin system.

This is how you would create an `invalid:` prefix that targets the `:invalid` pseudo class:

```javascript
// tailwind.config.js
const plugin = require('tailwindcss/plugin')

module.exports = {
  plugins: [
    plugin(function({ addVariant, e }) {
      addVariant('invalid', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return `.${e(`invalid${separator}${className}`)}:invalid`
        })
      })
    })
  ]
}
```

Before this has any effect, you need to enable the new `invalid` variant where you want to use it. 

Say you want to apply a different background color to an invalid element. You could add the `invalid` variant to the `backgroundColor` core plugin in your config file like so:

```javascript
// tailwind.config.js
const plugin = require('tailwindcss/plugin')

module.exports = {
  variants: {
    backgroundColor: ({after}) => after(['invalid'])
  },
  plugins: [
    plugin(function({ addVariant, e }) {
      addVariant('invalid', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return `.${e(`invalid${separator}${className}`)}:invalid`
        })
      })
    })
  ]
}
```

Now, you could make an input have a red background when invalid like so:

```html
<input type="text" class="bg-gray-100 invalid:bg-red-200" required />
```

Hope it helps!

**Reply by manix** [reply]
Yes, I noticed that it is documented fairly well. 

I thought it would be nice if they could be added without me having to do any modifications but as I continue expanding my knowledge on this subject I figure that using the cdn or core version defeats the purpose of the entire thing.

Thank you!

**Reply by simonswiss** [reply]
Yeah, it's tough to know what _should live in core_ vs what is _plugin territory_.

Worth knowing that every added variant can add a tremendous amount of generated classes, and therefore CSS output file size.

There is the `purge` option which takes care of things to an extent, but even in development mode, there is still a reason why all variants are not enabled by default.

I highly encourage you to look into working with Tailwind via a proper installation (PostCSS, config file etc) if you want to use it beyond a few projects where you pull the styles from the CDN. Happy to help you get a repo setup if you need! 👍

**Reply by manix** [reply]
Yep, I understand what you mean. 

It's kind of funny because the hobby project, or rather concept, I embarked on is based on building UI without processing. That's why I resisted introducing modifications and Tailwind out of the box seemed like the closest thing to modular css, without doing any processing.

I suppose a small sacrifice of the ideology will have to be made, but all in all is not so bad since Tailwind CLI is pretty easy to work with and best of all - stand alone.

Weirdly enough it's turning out pretty well, perhaps I might share it here when it's developed further.

**Reply by simonswiss** [reply]
Awesome - would love to see what you're working on with Tailwind!

**Reply by ashleyisles** [reply]
I had this exact same question! I've created this custom variant a few times now and was wondering if it would ever make it to core

**Reply by Zertz** [reply]
With the arrival of `@tailwindcss/jit`, would it make sense to reconsider adding this to core?

**Reply by jonathanstanley** [reply]
added in https://github.com/tailwindlabs/tailwindcss/pull/4482
